memstore/list/archive_vector_party.go

```diff
@@ -0,0 +1,367 @@
+	return &ArchiveVectorParty{
+		baseVectorParty: baseVectorParty{
+			length:   length,
+			dataType: dataType,
+		},
+		Pinnable: memstore.Pinnable{
+			AllUsersDone: sync.NewCond(locker),
+		},
+		totalValueBytes: totalValueBytes,
+	}
+}
+
+// Allocate allocate underlying storage for vector party. Note allocation for
+// archive vp does not report host memory change. Memory reporting is done
+// after switching to the new version of archive store. Before switching the memory
+// managed by this vp is counted as unmanaged memory.
+func (vp *ArchiveVectorParty) Allocate(hasCount bool) {
+	vp.offsets = memstore.NewVector(common.Uint32, vp.length*2)
+	vp.values = memstore.NewVector(common.Uint8, int(vp.totalValueBytes))
+}
+
+// GetBytes returns the bytes this vp occupies.
+func (vp *ArchiveVectorParty) GetBytes() int64 {
+	if vp.values != nil && vp.offsets != nil {
+		return int64(vp.values.Bytes + vp.offsets.Bytes)
+	}
+	return 0
+}
+
+// SafeDestruct destructs vector party memory.
+func (vp *ArchiveVectorParty) SafeDestruct() {
+	if vp != nil {
+		vp.offsets.SafeDestruct()
+		vp.offsets = nil
+		vp.values.SafeDestruct()
+		vp.values = nil
+	}
+}
+
+// AsList is the implementation from common.VectorParty
+func (vp *ArchiveVectorParty) AsList() common.ListVectorParty {
+	return vp
+}
+
+// Equals is the implementation from common.VectorParty
+func (vp *ArchiveVectorParty) Equals(other common.VectorParty) bool {
+	return vp.equals(other, vp.AsList())
+}
+
+// GetValue TODO handling invalid case
+func (vp *ArchiveVectorParty) getValue(row int) (val unsafe.Pointer, validity bool) {
+	offset, length := vp.GetOffsetLength(row)
+	if offset == 0 && length == 0 {
+		return unsafe.Pointer(uintptr(0)), false
+	}
+	baseAddr := uintptr(vp.values.Buffer())
+	return unsafe.Pointer(baseAddr + uintptr(offset)), true
+}
+
+// GetDataValue is not implemented in baseVectorParty
+func (vp *ArchiveVectorParty) GetDataValue(row int) common.DataValue {
+	val, valid := vp.getValue(row)
+	return common.DataValue{
+		DataType: vp.dataType,
+		Valid:    valid,
+		OtherVal: val,
+	}
+}
+
+// GetDataValueByRow just call GetDataValue
+func (vp *ArchiveVectorParty) GetDataValueByRow(row int) common.DataValue {
+	return vp.GetDataValue(row)
+}
+
+func (vp *ArchiveVectorParty) setValue(row int, val unsafe.Pointer, valid bool) {
+	if !valid {
+		var zero uint32
+		vp.offsets.SetValue(2*row, unsafe.Pointer(&zero))
+		vp.offsets.SetValue(2*row+1, unsafe.Pointer(&zero))
+		return
+	}
+	newLen := int(*(*uint32)(val))
+	newBytes := common.CalculateListElementBytes(vp.dataType, newLen)
+	// Set offset.
+	vp.offsets.SetValue(2*row, unsafe.Pointer(&vp.bytesWritten))
+	// Set length.
+	vp.offsets.SetValue(2*row+1, unsafe.Pointer(&newLen))
+
+	baseAddr := uintptr(vp.values.Buffer()) + uintptr(vp.bytesWritten)
+	to := cgoutils.MakeSliceFromCPtr(baseAddr, newBytes)
+	from := cgoutils.MakeSliceFromCPtr(uintptr(val), newBytes)
```

Comment:
caution in using MakeSliceFromCPtr, the pointer has to be C allocated only. or the underlying address might be changed by golang garbage collector

use our own memcopy can achieve the same